docs/reference/pip_install.rst

```diff
@@ -36,6 +39,33 @@ which depends on foo::
     ...
     Installing collected packages foo, baz, bar
 
+
+Prior to v6.1.0, pip made no commitments about install order.
+
+The decision to install topologically is based on the principle that
+installations should proceed in a way that leaves the environment usable at each
+step. This has two main practical benefits:
+
+1. Concurrent use of the environment during the install is more likely to work.
+2. A failed install won't leave broken packages.  Although pip would like to
```

Comment:
Failures that may leave a broken environment:
 - power failure [our fs operations are non-atomic and lack barriers]
 - Wheels may have been built but be broken.
 - A package may have incorrect runtime deps (e.g. missing a version cap or some such)
 - wheels may not be buildable and the package is missing things needed to build [that setup_requires can't describe]
 - permissions errors leading to installation failing with part of the thing installed
 - interrupts while the environment is in an inconsistent state [e.g. when upgrading A to A' we uninstall A then install A', so if we're interrupted between these anything depending on A is broken after we've terminated.

There are likely more - thats just offhand. E.g. many ways to break.

Topological sorting removes:
 - some cases of bad setup deps breaking builds
 - the amount of operations that take place with the environment in an inconsistent state, and so increases the safety of interrupting pip [whether the interruption is a failed build, operating system signals (out of memory etc) or users hitting ctrl-C or ctrl-\.